build: align build system with vue 3

kiaking opened the pull request:
This PR updates the build system for Vuex 4 to align with Vue 3.

- Bundles are now aligned with Vue 3 such as `vuex.esm-bundler.js` and `vuex.global.js`.
- Using `__DEV__` global variable and rollup replace plugin to handle `process.env.NODE_ENV !== 'production'` replacement.
- Remove `dist` folder from commit expect for `logger.d.ts`, plus organize `.gitignore` file to match the latest repository structure.
- Make `index.js` to be the primary esm entry point by renaming `index.esm.js` to `index.js`, and renaming `index.js` to `index.cjs`.
- Use index file without the build in test cases.
- ~⚠️ The "release script" should be updated once #1707 is merged in `dev` branch.~ Included the "release script" as well.

ktsn commented on .gitignore:
```diff
@@ -1,12 +1,10 @@
-types/typings
-types/test/*.js
-explorations
+/dist
```
I guess we need to add `!/dist/logger.d.ts` ?

kiaking replied:
Ah, good point! Yeah I'll add that. `logger.d.ts` will be tracked regardless of it because it is already committed, but it's nice to have precise definition 👍

kiaking replied:
And fixed 👍

kiaking commented:
@ktsn Please hold on merging this one. I'll include the changes from #1707 to this PR as well.

kiaking commented:
@ktsn OK, I think it's done. I've added `next` tag to the release script since `4.0` should be always release as `next`.

kiaking commented:
Hah, not sure why e2e test is failing... maybe running again would solve it...?

kiaking commented:
Added `jsdelivr` support, and renamed `esm` module to `esm-bundler` to align with the latest [Vue 3 commit](https://github.com/vuejs/vue-next/commit/218e6e166784ce967244c68c9a25822ad2053e11).

kiaking commented:
Hmmm... I still don't know why E2E test is failing on CI. It works locally... 🤔

kiaking commented:
YES!!! The test has passed. @ktsn If you were to merge this, please don't forget to squash the commits 😅
